mkosi/gentoo.py

```diff
@@ -0,0 +1,510 @@
+import multiprocessing
+import os
+import re
+import tarfile
+import urllib.parse
+import urllib.request
+from textwrap import dedent
+from typing import List, Set
+
+from .backend import (
+    ARG_DEBUG,
+    CommandLineArguments,
+    MkosiPrinter,
+    OutputFormat,
+    die,
+    run_workspace_command,
+)
+
+NO_PORTAGE_DIE_MSG = "You need portage(5) - the heart of Gentoo"
+PORTAGE_GIT_URL = "https://gitweb.gentoo.org/proj/portage.git"
+try:
+    from portage.const import (  # type: ignore
+        CUSTOM_PROFILE_PATH,
+        EBUILD_SH_ENV_DIR,
+        PROFILE_PATH,
+        USER_CONFIG_PATH,
+    )
+    from portage.package.ebuild.config import config  # type: ignore
+except ImportError:
+    die(f"{NO_PORTAGE_DIE_MSG} : {PORTAGE_GIT_URL}")
+
+
+# https://github.com/python/mypy/issues/5732
+kimg_path: str
+
+
+class Gentoo:
+    arch_profile: str
+    emerge_default_opts: Set[str]
+    EMERGE_UPDATE_OPTS: Set[str]
+    arch: str
+    package_use: str
+    pkgs_boot: List[Set[str]]
+    pkgs_sys: Set[str]
+    pkgs_fs: Set[str]
+    grub_platforms: Set[str]
+    portage_features: Set[str]
+    portage_use_flags: Set[str]
+    UNINSTALL_IGNORE: Set[str]
+    dracut_atom = "sys-kernel/dracut"
+
+    def __init__(self, args: CommandLineArguments, root: str, do_run_build_script: bool) -> None:
+        self.portage_cfg = config(config_root=root, target_root=root, sysroot=root, eprefix=None)
+        if "build-script" in ARG_DEBUG:
+            for c in self.portage_cfg:
+                print("%-32.32s = %-s" % (c, self.portage_cfg[c]))
+
+        jobs = multiprocessing.cpu_count()
+        self.emerge_default_opts = {
+            "--buildpkg=y",
+            "--usepkg=y",
+            "--keep-going=y",
+            "--jobs=" + str(jobs),
+            "--load-average=" + str(jobs - 1),
+            "--nospinner",
+        }
+        if "build-script" in ARG_DEBUG:
+            self.emerge_default_opts |= {"--verbose"}
```

Comment:
Hmm, so sets do not keep order. I think this should use a list, like everywhere else in the code. If you change `{}` to `[]`, and `|=` to `+=`, it should "just work".